src/main/java/com/epam/aidial/core/service/FileService.java

```diff
@@ -0,0 +1,173 @@
+                try (InputStream inputStream = payload.openStream()) {
+                    CacheService.Item<byte[]> item = new CacheService.Item<>(
+                            CacheService.ItemMetadata.builder()
+                                    .etag(etag)
+                                    .contentType(contentType)
+                                    .contentLength(length)
+                                    .build(),
+                            inputStream.readAllBytes());
+                    cacheService.cacheBytes(key, item);
+                    return FileStream.fromCacheItem(item);
+                }
+            }
+
+            return new FileStream(payload.openStream(), etag, contentType, length);
+        }
+    }
+
+    @Nullable
+    public String getEtag(ResourceDescription resource) {
+        FileMetadata metadata = getMetadata(resource);
+        if (metadata == null) {
+            return null;
+        }
+
+        return metadata.getEtag();
+    }
+
+    public FileMetadata getMetadata(ResourceDescription resource) {
+        String key = cacheService.redisKey(resource);
+        CacheService.Result<CacheService.ItemMetadata> metadata = cacheService.getMetadata(key);
+        if (metadata != null) {
+            return fileMetadataFromCacheItemMetadata(resource, metadata);
+        }
+
+        BlobMetadata blobMetadata = blobStorage.meta(resource.getAbsoluteFilePath());
+        if (blobMetadata == null) {
+            return null;
+        }
+
+        String contentType = blobMetadata.getContentMetadata().getContentType();
+        long contentLength = blobMetadata.getContentMetadata().getContentLength();
+        String etag = ResourceUtil.extractEtag(blobMetadata.getUserMetadata());
+
+        return (FileMetadata) new FileMetadata(resource, contentLength, contentType).setEtag(etag);
+    }
+
+    public void putFile(
+            ResourceDescription resource, byte[] bytes, String contentType, String newEtag, EtagHeader etag) {
+        String key = cacheService.redisKey(resource);
+        try (LockService.Lock ignored = lockService.lock(key)) {
+            FileMetadata metadata = getMetadata(resource);
+            if (metadata != null) {
+                etag.validate(metadata.getEtag());
+            }
+
+            CacheService.ItemMetadata newMetadata = CacheService.ItemMetadata.builder()
+                    .etag(newEtag)
+                    .contentType(contentType)
+                    .contentLength((long) bytes.length)
+                    .build();
+            CacheService.Item<byte[]> item = new CacheService.Item<>(newMetadata, bytes);
+            cacheService.saveBytes(key, item);
+            if (metadata == null) {
+                // create an empty object for listing
+                cacheService.saveStub(key, newMetadata);
+            }
+        }
+    }
+
+    public void putFile(ResourceDescription resource, MultipartUpload mpu, List<MultipartPart> parts, EtagHeader etag) {
```

Comment:
I don't think this flows follows the diagram you provided. I see no lock refresh during chunk/part upload. Why etag validation happens after all parts uploaded? why cache service evolved for big files?

Reply:
Yes, we discussed the diagrams with @artsiomkorzun . We agreed to remove "lock 1". For multipart upload, the ETag check occurs before the upload outside of a lock and once again under "lock 2". I'll update the diagrams.